### PR TITLE
Replace DeltaNet intra-chunk scan with closed-form chunked delta rule

### PR DESCRIPTION
**Pull request by ry2009**
## What

The chunked DeltaNet prefill path computed each chunk's intra-chunk result with a
per-token `jax.lax.scan` -- `chunk_size` sequential steps per chunk, each doing tiny
per-head work. That sequential dependency is latency-bound and a large share of
prefill time on GPU.

This replaces it with the **closed-form chunked delta rule** (UT-transform / WY
representation): each chunk's `chunk_outputs`, `correction_vecs`, `end_state`, and
`end_prop` come from dense fp32 matmuls plus unit-lower triangular solves per
(chunk, head). Same recurrence, no per-token intra-chunk scan.

- `lalamo/modules/token_mixers/chunked_delta.py` -- the kernel. It has no lalamo deps,
  casts runtime inputs to fp32 once, uses fp32 matmul accumulation, and runs chunks
  with `jax.lax.map` so the dense `chunk_size x chunk_size` intermediates stay scoped
  to one chunk. Heads are still vmapped.
- `deltanet.py` -- calls `chunk_delta_forward` in `_chunked_scan`; the now-dead
  per-token scan (`_intra_chunk_token_step` / `_intra_chunk_scan`) and its helper
  NamedTuples are removed. `ChunkKernelResult` is a field-for-field match for the old
  `DeltaNetChunkScanResult`, so the inter-chunk fold-in downstream is unchanged.
- `tests/unit/test_chunked_delta.py` -- validates the kernel against an fp64 per-token
  oracle across chunk sizes {32, 64, 128}, including `key_channels != value_channels`.
  `tests/unit/modules/test_deltanet.py` compares the full chunked scan against the
  recurrent scan across tail/chunk configurations.

## Correctness

The closed form *is* the per-token recurrence, re-associated. The current kernel is
gated against the fp64 oracle with `rtol=1e-4, atol=1e-5`; the remaining error is fp32
roundoff from the production math path.

Because it computes a different *order* of operations than the sequential scan, it is
numerically equivalent within fp32 roundoff (~1e-4) but **not bit-identical**.
Validated on Qwen3.6-27B (deterministic ops):

- **Teacher-forced top-1 agreement vs the scan: 99.2%** over 2517 positions, and the
  disagreements occur **only at near-ties** -- the median scan top1-minus-top2 margin
  at a disagreement is ~0.02 vs ~1.72 typical (~100x smaller).
- **GSM8K exact-match (N=80, greedy): 97.5% (kernel) vs 97.5% (scan)** -- identical;
  the 2 differing problems flip in both directions (net zero).

So the swap is quality-neutral. Note that full *free-running* greedy decoding is not
bit-identical between any closed-form kernel and a sequential scan (a ~1e-4 roundoff
flips near-tie argmaxes and the continuations diverge) -- which is why equivalence is
measured by teacher-forced top-1 agreement + task accuracy, not token-sequence
identity.

## Effect

Original GPU benchmark:

- 512-token prefill (batch 1) is **~1.36x faster** than the per-token scan.
- The intra-chunk computation itself is **~7x faster** (batched matmuls/triangular
  solves vs a sequential scan over `chunk_size` tokens).

After the cleanup pass, a local CPU regression sanity check of the current branch
against the old scan on a 512-token shaped kernel input (`16 x 32`, `heads=4`,
`key_channels=96`, `value_channels=128`) still shows a speedup:

- old scan: `0.006341s`
- current closed form: `0.004323s`
- speedup: **1.47x**

That CPU number is only a regression guard; the headline effect remains the GPU
benchmark above.

### CODE REVIEW

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+"""Closed-form DeltaNet intra-chunk recurrence."""
+
+from __future__ import annotations
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: we don't use future annotations and useless comments too

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    left: Array,
+    right: Array,
+    precision: jax.lax.Precision | None = None,
+) -> Array:
+    return jnp.matmul(
+        left,
+        right,
+        preferred_element_type=jnp.float32,
+        precision=precision,
+    )
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: somewhere elsewhere in codebase we use`with jax.default_matmul_precision`; I don't fully understand why can't you just use that?

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def _solve_unit_lower(matrix: Array, rhs: Array) -> Array:
+    return jax.scipy.linalg.solve_triangular(
+        matrix,
+        rhs,
+        lower=True,
+        unit_diagonal=True,
+    )
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: inline

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    values: Float[Array, "chunk_size value_channels"],
+    decay_factor: Array,
+    beta: Array,
+) -> tuple[
+    Float[Array, "chunk_size value_channels"],
+    Float[Array, "chunk_size key_channels"],
+    Float[Array, "value_channels key_channels"],
+    Float[Array, "key_channels key_channels"],
+]:
+    """Closed-form intra-chunk pass for one head: (outputs, correction_vecs, end_state, end_prop)."""
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: not sure how useful the comment is

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    chunk_size, key_channels = keys.shape
+    beta = beta.astype(jnp.float32)
+    queries = queries.astype(jnp.float32)
+    keys = keys.astype(jnp.float32)
+    values = values.astype(jnp.float32)
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py:35]: this is actually slightly dangerous, just to make sure we are aligned:

i spent a lot of effort a month ago matching uzu/lalamo precisions inside of the deltanet; if this preserves the old precisions - good, if it doesn't - bad

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    cumulative_decay = jnp.cumsum(decay_factor.astype(jnp.float32), axis=0)
+    final_decay = cumulative_decay[-1]
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: maybe unpacking? *_, final_decay = ...?

File: lalamo/modules/token_mixers/chunked_delta.py (hunk 2 above)
Comment (knyazer) [reply lalamo/modules/token_mixers/chunked_delta.py]: i'm also not fully certain what's the point - everywhere i see you just use this thing with default params ,which is equivalent to default matmuls, no?

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        chunk_inputs: tuple[
+            Float[Array, "chunk_size heads key_channels"],
+            Float[Array, "chunk_size heads key_channels"],
+            Float[Array, "chunk_size heads value_channels"],
+            Float[Array, "chunk_size heads"],
+            Float[Array, "chunk_size heads"],
+        ],
+    ) -> tuple[Array, Array, Array, Array]:
+        chunk_queries, chunk_keys, chunk_values, chunk_decay, chunk_beta = chunk_inputs
+        return jax.vmap(_single_head, in_axes=(1, 1, 1, 1, 1))(
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: can't you just jax.vmap(_single_head, in_axes=1)(chunk_inputs)? then you might even be able to just inline this function into the vmap, like jax.vmap(jax.vmap(_single_head, in_axes=1), in_axes=0)((queries, keys, values, ...))

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            chunk_outputs,
+            "num_chunks heads chunk_size value_channels -> num_chunks chunk_size heads value_channels",
+        ).astype(jnp.float32),
+        einops.rearrange(
+            correction_vecs,
+            "num_chunks heads chunk_size key_channels -> num_chunks chunk_size heads key_channels",
+        ).astype(jnp.float32),
+        end_state.astype(jnp.float32),
+        end_prop.astype(jnp.float32),
+    )
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py:92]: the casts are unnecessasry?

File: tests/unit/test_chunked_delta.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
```
Comment (knyazer) [inline tests/unit/test_chunked_delta.py]: no future

File: tests/unit/test_chunked_delta.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+@pytest.fixture
+def enable_x64() -> Iterator[None]:
+    x64_was_enabled = jax.config.read("jax_enable_x64")
+    jax.config.update("jax_enable_x64", val=True)
+    try:
+        yield
+    finally:
+        jax.config.update("jax_enable_x64", val=x64_was_enabled)
+
```
Comment (knyazer) [inline tests/unit/test_chunked_delta.py:20]: i'm fairly confident that this is not a good idea, mainly because I don't think it would work(?)

i always thought that config-setting during runtime is undefined behaviour; like as soon as you instantiated a jax array doing any config switching is dangerous imo.

ultimately, i'm glad it works, i don't necessarily see how is this necessary lol

otherwise tests are nice, good job

File: lalamo/modules/token_mixers/chunked_delta.py (hunk 5 above)
Comment (ry2009) [reply lalamo/modules/token_mixers/chunked_delta.py:35]: ran the old _chunked_scan (fp32) directly against the new closed-form on identical inputs. on CPU (true fp32) they match to ~1e-6 on every output, so it's the same computation just round-off. on GPU it's ~5e-4 on the attn outputs (plain TF32, same as the old einsums) and ~1.5e-5 on end_state. the F32_F32_F32 on end_state is on purpose: the old scan accumulated the state elementwise in true fp32, so true-fp32 here is exactly what keeps it matching the old path (default/TF32 there drifts ~6e-4). deltanet-level fp32 casts are unchanged, so precision is preserved but we can review this more in depth if wanted

File: lalamo/modules/token_mixers/chunked_delta.py (hunk 6 above)
Comment (ry2009) [reply lalamo/modules/token_mixers/chunked_delta.py]: keeping cumulative_decay[-1] here — *_, final_decay = iterates the array, which unrolls into chunk_size slice ops under jit, whereas [-1] is a single gather. same value, cheaper.

File: tests/unit/test_chunked_delta.py (hunk 10 above)
Comment (ry2009) [reply tests/unit/test_chunked_delta.py:20]: it does work bc it's the standard jax-test pattern: flip jax_enable_x64 before any array exists and restore in finally so it stays scoped and doesn't leak. fp64 is load-bearing for the oracle (in fp32 you can't separate a real bug from round-off ... the gpu path sits ~5e-4, well above a fp32 oracle's noise) but can swap to a different x64 hook if there's a preferred one and thanks re: the tests 🙏

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
+"""Closed-form DeltaNet intra-chunk recurrence."""
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: comment kill

File: lalamo/modules/token_mixers/chunked_delta.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        end_state = jnp.matmul(update_values.T, keys * end_state_weights[:, None])
+
+    prop_matrix = eye_chunk + jnp.tril(beta[:, None] * key_dot_key, -1)
+    prop_updates = jax.scipy.linalg.solve_triangular(
+        prop_matrix, beta[:, None] * keys, lower=True, unit_diagonal=True
+    )
+
+    key_dot_query = jnp.matmul(keys, queries.T)
+    past_key_dot_query = jnp.triu(key_dot_query, 0)
+    correction = queries - jnp.matmul(past_key_dot_query.T, prop_updates)
```
Comment (knyazer) [inline lalamo/modules/token_mixers/chunked_delta.py]: jnp.matmul == @ - please use the dog operator

File: tests/unit/test_chunked_delta.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        chunk_decay: Array,
+        chunk_beta: Array,
+    ) -> tuple[Array, Array, Array, Array]:
+        return jax.vmap(_reference_single_head, in_axes=(1, 1, 1, 1, 1))(
+            chunk_queries,
+            chunk_keys,
+            chunk_values,
+            chunk_decay,
+            chunk_beta,
+        )
```
Comment (knyazer) [inline tests/unit/test_chunked_delta.py]: same thing with inlining vmap applies here

File: tests/unit/test_chunked_delta.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    beta = jax.nn.sigmoid(jax.random.normal(beta_key, (num_chunks, chunk_size, num_heads), dtype=jnp.float64))
+    decay_factor = -jax.nn.softplus(
+        jax.random.normal(decay_key, (num_chunks, chunk_size, num_heads), dtype=jnp.float64)
+    )
+    return queries, keys, values, decay_factor, beta
+
+
+def _assert_close(result: Array, reference: Array) -> None:
+    assert_close(result=result, reference=reference.astype(jnp.float32), rtol=1e-4, atol=1e-5)
+
```
Comment (knyazer) [inline tests/unit/test_chunked_delta.py]: inline?

File: lalamo/modules/token_mixers/chunked_delta.py (hunk 6 above)
Comment (knyazer) [reply lalamo/modules/token_mixers/chunked_delta.py]: image


hm?

Comment (knyazer) [reply lalamo/modules/token_mixers/chunked_delta.py]: oh maybe i'm testing on my mac not cuda, one sec

Comment (knyazer) [reply lalamo/modules/token_mixers/chunked_delta.py]: okay i will never believe anything you say from now on 
image

File: lalamo/modules/token_mixers/chunked_delta.py (hunk 12 above)
Comment (knyazer) [reply lalamo/modules/token_mixers/chunked_delta.py]: (or "at" operator, whatever)

**Comment by knyazer** [comment]
also fix thge linter stuffs please

File: lalamo/modules/token_mixers/chunked_delta.py (hunk 6 above)
Comment (ry2009) [reply lalamo/modules/token_mixers/chunked_delta.py]: i meant jaxpr ;(

Comment (ry2009) [reply lalamo/modules/token_mixers/chunked_delta.py]: ha, you got me — checked the lowered output and XLA folds the unpack back to a single stablehlo.slice, so the compiled HLO is identical. i was reading the jaxpr, not the lowered hlo. switched to *_, final_decay 🤝

**Review by knyazer — APPROVED** [review]
lgtm
